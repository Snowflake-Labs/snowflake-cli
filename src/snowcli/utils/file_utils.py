from __future__ import annotations

import os
import shutil
from pathlib import Path

from jinja2 import Environment, FileSystemLoader

templates_path = Path(__file__).parent / "../python_templates"


def prepare_app_zip(file_path: Path, temp_dir: str) -> str:
    # get filename from file path (e.g. app.zip from /path/to/app.zip)
    # TODO: think if no file exceptions are handled correctly
    file_name = file_path.name
    temp_path = temp_dir + "/" + file_name
    shutil.copy(file_path, temp_path)
    return temp_path


def generate_snowpark_coverage_wrapper(
    target_file: str,
    proc_name: str,
    proc_signature: str,
    handler_module: str,
    handler_function: str,
    coverage_reports_stage_path: str,
) -> None:
    """Using a hardcoded template (python_templates/snowpark_coverage.py.jinja), substitutes variables
    and writes out a file.
    The resulting file can be used as the initial handler for the stored proc, and uses the coverage package
    to measure code coverage of the actual stored proc code.
    Afterwards, the handler persists the coverage report to json by executing a query.

    Args:
        target_file (str): _description_
        proc_name (str): _description_
        proc_signature (str): _description_
        handler_module (str): _description_
        handler_function (str): _description_
    """

    environment = Environment(loader=FileSystemLoader(templates_path))
    template = environment.get_template("snowpark_coverage.py.jinja")
    content = template.render(
        {
            "proc_name": proc_name,
            "proc_signature": proc_signature,
            "handler_module": handler_module,
            "handler_function": handler_function,
            "coverage_reports_stage_path": coverage_reports_stage_path,
        }
    )
    with open(target_file, "w", encoding="utf-8") as output_file:
        output_file.write(content)


def create_project_template(
    template_name: str, project_directory: str | None = None
) -> None:
    target = project_directory or os.getcwd()
    shutil.copytree(
        Path(importlib.util.find_spec("templates").origin).parent / template_name,  # type: ignore
        target,
        dirs_exist_ok=True,
    )
