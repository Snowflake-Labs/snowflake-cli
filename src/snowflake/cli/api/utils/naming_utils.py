from __future__ import annotations

import re
from typing import Tuple

from snowflake.cli.api.project.util import (
    VALID_IDENTIFIER_REGEX,
)


def from_qualified_name(name: str) -> Tuple[str, str | None, str | None]:
    """
    Takes in an object name in the form [[database.]schema.]name. Returns a tuple (name, [schema], [database])
    """
    # TODO: Use regex to match object name to a valid identifier or valid identifier (args). Second case is for sprocs and UDFs
    qualifier_pattern = rf"(?:(?P<first_qualifier>{VALID_IDENTIFIER_REGEX})\.)?(?:(?P<second_qualifier>{VALID_IDENTIFIER_REGEX})\.)?(?P<name>.*)"
    result = re.fullmatch(qualifier_pattern, name)

    if result is None:
        raise ValueError(f"'{name}' is not a valid qualified name")

    unqualified_name = result.group("name")
    if result.group("second_qualifier") is not None:
        database = result.group("first_qualifier")
        schema = result.group("second_qualifier")
    else:
        database = None
        schema = result.group("first_qualifier")
    return unqualified_name, schema, database
