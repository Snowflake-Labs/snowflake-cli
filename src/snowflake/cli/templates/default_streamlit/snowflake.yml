definition_version: "2"
entities:
  my_streamlit:
    type: "streamlit"
    name: "my_dashboard"
    title: "My Streamlit Dashboard"
    stage: "my_streamlit_stage"
    query_warehouse: 'my_streamlit_warehouse'
    main_file: streamlit_app.py
    pages_dir: pages/
    artifacts:
      - streamlit_app.py
      - environment.yml
      - common/hello.py
      - pages/
