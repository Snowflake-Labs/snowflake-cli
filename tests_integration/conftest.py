from __future__ import annotations

import functools
import pytest
import toml

from pathlib import Path
from snowcli.cli import app
from tempfile import NamedTemporaryFile
from typer import Typer
from typer.testing import CliRunner

TEST_DIR = Path(__file__).parent


@pytest.fixture(scope="session")
def test_snowcli_config():
    test_config = TEST_DIR / "config/test.toml"
    config = toml.load(test_config)
    config["snowsql_config_path"] = str(TEST_DIR / "config/connection_configs.toml")
    with NamedTemporaryFile(suffix=".toml", mode="w+") as fh:
        toml.dump(config, fh)
        fh.flush()
        yield fh.name


class SnowCLIRunner(CliRunner):
    def __init__(self, app: Typer, test_snowcli_config: str):
        super().__init__()
        self.app = app
        self.test_snowcli_config = test_snowcli_config

    @functools.wraps(CliRunner.invoke)
    def invoke(self, *a, **kw):
        return super().invoke(self.app, *a, **kw)

    def invoke_with_config(self, *args, **kwargs):
        return self.invoke(
            ["--config-file", self.test_snowcli_config, *args[0]], **kwargs
        )


@pytest.fixture
def runner(test_snowcli_config):
    return SnowCLIRunner(app, test_snowcli_config)
