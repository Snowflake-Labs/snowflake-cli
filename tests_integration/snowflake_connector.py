from __future__ import annotations

import os
import pytest
import uuid

from snowflake import connector
from snowcli.exception import EnvironmentVariableNotFoundError

_ENV_PARAMETER_PREFIX = "SNOWFLAKE_CONNECTIONS_INTEGRATION"


@pytest.fixture(scope="session")
def create_database():
    database_name = f"db_{uuid.uuid4().hex}"

    config = {
        "application": "INTEGRATION_TEST",
        "account": _get_from_env("ACCOUNT", True),
        "user": _get_from_env("USER", True),
        "password": _get_from_env("PASSWORD", True),
        "schema": _get_from_env("SCHEMA", False, "PUBLIC"),
        "host": _get_from_env("HOST", False),
        "role": _get_from_env("ROLE", False),
    }
    config = {k: v for k, v in config.items() if v is not None}
    connection = connector.connect(**config)

    connection.execute_string(f"create database {database_name}")
    os.environ[f"{_ENV_PARAMETER_PREFIX}_DATABASE"] = database_name

    yield database_name

    connection.execute_string(f"drop database {database_name}")
    del os.environ[f"{_ENV_PARAMETER_PREFIX}_DATABASE"]


@pytest.fixture(scope="session")
def snowflake_session(create_database):
    config = {
        "application": "INTEGRATION_TEST",
        "account": _get_from_env("ACCOUNT", True),
        "user": _get_from_env("USER", True),
        "password": _get_from_env("PASSWORD", True),
        "host": _get_from_env("HOST", False),
        "database": _get_from_env("DATABASE", False),
        "role": _get_from_env("ROLE", False),
        "schema": "PUBLIC",
        "warehouse": _get_from_env("WAREHOUSE", False),
    }
    config = {k: v for k, v in config.items() if v is not None}
    connection = connector.connect(**config)
    yield connection
    connection.close()


def _get_from_env(parameter_name: str, required: bool, default=None) -> str | None:
    env_value = os.environ.get(f"{_ENV_PARAMETER_PREFIX}_{parameter_name}")
    if required and not env_value:
        if default is None:
            raise EnvironmentVariableNotFoundError(
                f"{_ENV_PARAMETER_PREFIX}_{parameter_name}"
            ) from None
        return default
    return env_value
