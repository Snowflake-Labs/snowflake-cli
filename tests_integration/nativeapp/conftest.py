from contextlib import contextmanager

import pytest

from tests_integration.conftest import SnowCLIRunner


@pytest.fixture
def nativeapp_project_directory(project_directory, nativeapp_teardown):
    @contextmanager
    def _nativeapp_project_directory(name):
        with project_directory(name) as d:
            with nativeapp_teardown():
                yield d

    return _nativeapp_project_directory


@pytest.fixture
def nativeapp_teardown(runner: SnowCLIRunner):
    @contextmanager
    def _nativeapp_teardown():
        try:
            yield
        finally:
            result = runner.invoke_with_connection_json(
                ["app", "teardown", "--force", "--cascade"]
            )
            assert result.exit_code == 0

    return _nativeapp_teardown
