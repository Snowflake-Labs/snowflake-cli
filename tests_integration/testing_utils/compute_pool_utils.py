import math
import time

import pytest
from snowflake.connector import SnowflakeConnection

from tests_integration.conftest import SnowCLIRunner
from tests_integration.test_utils import contains_row_with, not_contains_row_with
from tests_integration.testing_utils.assertions.test_result_assertions import (
    assert_that_result_is_successful_and_executed_successfully,
)


class ComputePoolTestSetup:
    def __init__(
        self,
        runner: SnowCLIRunner,
        snowflake_session: SnowflakeConnection,
    ):
        self.runner = runner
        self.snowflake_session = snowflake_session


class ComputePoolTestSteps:
    def __init__(self, setup: ComputePoolTestSetup):
        self._setup = setup

    def create_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "compute-pool",
                "create",
                compute_pool_name,
                "--min-nodes",
                1,
                "--family",
                "CPU_X64_XS",
            ]
        )
        assert result.json, result.output
        assert "status" in result.json
        assert (
            f"Compute Pool {compute_pool_name.upper()} successfully created."
            in result.json["status"]  # type: ignore
        )

    def list_should_return_compute_pool(self, compute_pool_name) -> None:
        result = self._execute_list()
        assert contains_row_with(result.json, {"name": compute_pool_name.upper()})

    def list_should_not_return_compute_pool(self, compute_pool_name: str) -> None:
        result = self._execute_list()
        assert not_contains_row_with(result.json, {"name": compute_pool_name.upper()})

    def describe_should_return_compute_pool(self, compute_pool_name: str) -> None:
        result = self._execute_describe(compute_pool_name)
        assert result.json
        assert len(result.json) == 1
        assert result.json[0]["name"] == compute_pool_name.upper()

    def drop_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "drop",
                "compute-pool",
                compute_pool_name,
            ],
        )
        assert result.json
        assert len(result.json) == 1
        assert result.json[0] == {  # type: ignore
            "status": f"{compute_pool_name.upper()} successfully dropped."
        }

    def stop_all_on_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "stop-all", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def suspend_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "suspend", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def resume_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "resume", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def wait_until_compute_pool_is_idle(self, compute_pool_name: str) -> None:
        self._wait_until_compute_pool_reaches_state(compute_pool_name, "IDLE", 300)

    def wait_until_compute_pool_is_suspended(self, compute_pool_name: str) -> None:
        self._wait_until_compute_pool_reaches_state(compute_pool_name, "SUSPENDED", 60)

    def _wait_until_compute_pool_reaches_state(
        self, compute_pool_name: str, target_state: str, max_duration: int
    ):
        assert max_duration > 0
        max_counter = math.ceil(max_duration / 10)
        target_state = target_state.upper()
        for i in range(max_counter):
            status = self._execute_describe(compute_pool_name)
            if contains_row_with(status.json, {"state": target_state}):
                return
            time.sleep(10)
        status = self._execute_describe(compute_pool_name)

        error_message = f"Compute pool {compute_pool_name} didn't reach target state '{target_state}' in {max_duration} seconds. Current state is '{status.json['state']}'"
        pytest.fail(error_message)

    def _execute_describe(self, compute_pool_name: str):
        return self._setup.runner.invoke_with_connection_json(
            ["object", "describe", "compute-pool", compute_pool_name]
        )

    def _execute_list(self):
        return self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "list",
                "compute-pool",
            ],
        )
