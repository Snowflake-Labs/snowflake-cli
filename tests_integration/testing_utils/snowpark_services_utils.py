import json
import math
import time
from textwrap import dedent
from typing import Union

import pytest
from snowflake.connector import SnowflakeConnection

from tests_integration.conftest import SnowCLIRunner
from tests_integration.test_utils import contains_row_with, not_contains_row_with
from tests_integration.testing_utils.assertions.test_result_assertions import (
    assert_that_result_is_successful_and_output_json_contains,
    assert_that_result_is_successful_and_output_json_equals,
)


class SnowparkServicesTestSetup:
    def __init__(
        self,
        runner: SnowCLIRunner,
        snowflake_session: SnowflakeConnection,
        test_root_path,
    ):
        self.runner = runner
        self.snowflake_session = snowflake_session
        self.test_root_path = test_root_path


class SnowparkServicesTestSteps:
    compute_pool = "snowcli_compute_pool"
    database = "snowcli_db"
    schema = "public"
    container_name = "echo-test"

    def __init__(self, setup: SnowparkServicesTestSetup):
        self._setup = setup

    def create_service(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "service",
                "create",
                service_name,
                "--compute-pool",
                self.compute_pool,
                "--spec-path",
                self._get_spec_path(),
                "--database",
                self.database,
                "--schema",
                self.schema,
            ],
        )
        assert_that_result_is_successful_and_output_json_equals(
            result, {"status": f"Service {service_name.upper()} successfully created."}
        )

    def status_should_return_service(self, service_name: str) -> None:
        result = self._execute_status(service_name)
        assert_that_result_is_successful_and_output_json_contains(
            result,
            {"containerName": self.container_name, "serviceName": service_name.upper()},
        )

    def logs_should_return_service_logs(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection(
            [
                "spcs",
                "service",
                "logs",
                service_name,
                "--container-name",
                self.container_name,
                "--instance-id",
                "0",
                "--database",
                self.database,
                "--schema",
                self.schema,
            ],
        )
        assert result.output
        # Assert this instead of full payload due to log coloring
        assert service_name in result.output
        assert '"GET /healthcheck HTTP/1.1" 200 -' in result.output

    def list_should_return_service(self, service_name: str) -> None:
        result = self._execute_list()
        assert contains_row_with(result.json, {"name": service_name.upper()})

    def list_should_not_return_service(self, service_name: str) -> None:
        result = self._execute_list()
        assert not_contains_row_with(result.json, {"name": service_name.upper()})

    def describe_should_return_service(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "describe",
                "service",
                f"{self.database}.{self.schema}.{service_name}",
            ],
        )
        assert result.json
        assert result.json[0]["name"] == service_name.upper()  # type: ignore

    def drop_service(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "drop",
                "service",
                f"{self.database}.{self.schema}.{service_name}",
            ],
        )
        assert result.json[0] == {  # type: ignore
            "status": f"{service_name.upper()} successfully dropped."
        }

    def wait_until_service_is_ready(self, service_name: str) -> None:
        self._wait_until_service_reaches_state(service_name, "READY", 900)

    def _wait_until_service_reaches_state(
        self, service_name: str, target_status: Union[str, dict], max_duration: int
    ):
        assert max_duration > 0
        max_counter = math.ceil(max_duration / 10)
        if isinstance(target_status, str):
            target_status = {"status": target_status}
        for i in range(max_counter):
            status = self._execute_status(service_name)
            if contains_row_with(status.json, target_status):
                return
            elif contains_row_with(status.json, {"status": "FAILED"}):
                describe = self._setup.runner.invoke_with_connection_json(
                    ["object", "describe", "service", service_name]
                )
                pytest.fail(
                    dedent(
                        f"""
                    {service_name} service failed before reaching target state:
                    {json.dumps(target_status)}
                    current state:
                    {json.dumps(status)}
                    current describe:
                    {json.dumps(describe)}
                    """
                    )
                )
            time.sleep(10)
        status = self._execute_status(service_name)

        error_message = f"""
{service_name} service didn't reach target state in {max_duration} seconds.
target:
{json.dumps(target_status)}
current:
{json.dumps(status.json)}
"""
        pytest.fail(error_message)

    def _execute_status(self, service_name: str):
        return self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "service",
                "status",
                service_name,
                "--database",
                self.database,
                "--schema",
                self.schema,
            ],
        )

    def _execute_list(self):
        return self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "list",
                "service",
            ],
        )

    def _get_spec_path(self):
        return f"{self._setup.test_root_path}/spcs/spec/spec.yml"
