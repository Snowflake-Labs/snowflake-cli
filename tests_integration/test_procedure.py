from __future__ import annotations

import re
from typing import Any, List, Dict

import pytest
from snowflake.connector import SnowflakeConnection
from syrupy import SnapshotAssertion

from tests_integration.testing_utils.file_utils import replace_text_in_file
from tests_integration.conftest import runner
from tests_integration.snowflake_connector import create_database, snowflake_session
from tests_integration.testing_utils.snowpark_utils import SnowparkTestSteps, SnowparkTestSetup
from tests_integration.testing_utils.sql_utils import sql_test_helper
from tests_integration.testing_utils.naming_utils import object_name_provider
from tests_integration.testing_utils.working_directory_utils import (
    temporary_working_directory,
)

from tests_integration.conftest import SnowCLIRunner
from tests_integration.testing_utils.assertions.test_file_assertions import (
    assert_that_current_working_directory_contains_only_following_files,
    assert_that_file_content_is_equal_to_snapshot,
)
from tests_integration.testing_utils.assertions.test_result_assertions import (
    assert_that_result_is_successful_and_has_no_output,
    assert_that_result_is_successful_and_output_contains,
    assert_that_result_is_successful,
    assert_that_result_is_successful_and_json_output_contains_value_at_path,
)
from tests_integration.testing_utils.naming_utils import ObjectNameProvider
from tests_integration.testing_utils.sql_utils import SqlTestHelper


@pytest.mark.integration
def test_snowpark_procedure_flow(_test_steps):
    _test_steps.assert_that_no_procedures_are_in_snowflake()
    _test_steps.assert_that_no_files_are_staged_in_snowflake()

    _test_steps.snowpark_procedure_list_should_return_no_data()

    _test_steps.snowpark_procedure_init_should_initialize_files_with_default_content()
    _test_steps.snowpark_procedure_package_should_zip_files()

    procedure_name = _test_steps.snowpark_procedure_create_should_finish_successfully()
    _test_steps.assert_that_only_these_procedures_are_in_snowflake(
        f"{procedure_name}() RETURN VARCHAR"
    )
    _test_steps.assert_that_only_these_files_are_staged_in_snowflake(
        f"deployments/{procedure_name}/app.zip"
    )

    _test_steps.snowpark_procedures_list_should_return_list_containing_procedure_at_the_first_place(
        procedure_name=procedure_name,
        arguments="()",
        result_type="VARCHAR",
    )

    _test_steps.snowpark_procedure_describe_should_return_procedure_description(
        procedure_name=procedure_name,
        arguments="()",
    )

    _test_steps.snowpark_procedure_execute_should_return_expected_value(
        procedure_name=procedure_name,
        arguments="()",
        expected_value="Hello World!",
    )

    _test_steps.snowpark_procedure_update_should_finish_successfully(procedure_name)
    _test_steps.assert_that_only_these_procedure_are_in_snowflake(
        f"{procedure_name}() RETURN NUMBER"
    )
    _test_steps.assert_that_only_these_files_are_staged_in_snowflake(
        f"deployments/{procedure_name}/app.zip"
    )

    _test_steps.snowpark_procedure_list_should_return_list_containing_procedure_at_the_first_place(
        procedure_name=procedure_name,
        arguments="()",
        result_type="NUMBER",
    )

    _test_steps.snowpark_procedure_execute_should_return_expected_value(
        procedure_name=procedure_name,
        arguments="()",
        expected_value=1,
    )

    _test_steps.snowpark_procedure_drop_should_finish_successfully(
        procedure_name=procedure_name,
        arguments="()",
    )
    _test_steps.assert_that_no_procedures_are_in_snowflake()
    _test_steps.assert_that_only_these_files_are_staged_in_snowflake(
        f"deployments/{procedure_name}/app.zip"
    )

    _test_steps.snowpark_procedure_list_should_return_no_data()


@pytest.fixture
def _test_setup(
    runner,
    snowflake_session,
    sql_test_helper,
    object_name_provider,
    temporary_working_directory,
    snapshot,
):
    snowpark_procedure_test_setup = SnowparkTestSetup(
        runner=runner,
        snowflake_session=snowflake_session,
        sql_test_helper=sql_test_helper,
        object_name_provider=object_name_provider,
        snapshot=snapshot,
    )
    yield snowpark_procedure_test_setup
    snowpark_procedure_test_setup.clean_after_test_case()


@pytest.fixture
def _test_steps(_test_setup):
    yield SnowparkTestSteps(_test_setup)

