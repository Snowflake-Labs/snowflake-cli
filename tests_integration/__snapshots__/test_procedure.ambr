# serializer version: 1
# name: test_snowpark_procedure_flow[.gitignore]
  '''
  .packages/
  .venv/
  app.zip
  __pycache__
  
  '''
# ---
# name: test_snowpark_procedure_flow[app.py]
  '''
  from __future__ import annotations
  
  import sys
  
  from snowflake.snowpark import Session
  
  
  def hello(session: Session, name: str) -> str:
      return f"Hello {name}"
  
  
  def test(session: Session) -> str:
      return "Test procedure"
  
  
  # For local debugging. Be aware you may need to type-convert arguments if
  # you add input parameters
  if __name__ == "__main__":
      from snowcli.config import cli_config
  
      session = Session.builder.configs(cli_config.get_connection("dev")).create()
      if len(sys.argv) > 1:
          print(hello(session, *sys.argv[1:]))  # type: ignore
      else:
          print(hello(session))  # type: ignore
      session.close()
  
  '''
# ---
# name: test_snowpark_procedure_flow[requirements.txt]
  '''
  snowflake-snowpark-python
  #snowcli  # for local development
  
  '''
# ---
# name: test_snowpark_procedure_flow[snowflake.yml]
  '''
  definition_version: 1
  procedures:
    - name: hello
      handler: "app.hello"
      signature:
        - name: "name"
          type: "string"
      returns: string
    - name: test
      handler: "app.test"
      signature: ""
      returns: string
  
  '''
# ---
